### ISSUE TITLE
Repeatedly writing to a `Proc` eats all available memory

### ISSUE THREAD

**Issue by adamsmd** [rakudo/rakudo#3858]
## The Problem

Repeatedly writing to a `Proc` eats all available memory

## Expected Behavior

Memory usage remains constant

## Actual Behavior

Memory usage continually grows

## Steps to Reproduce

The following program demonstrates this problem.  The longer this program runs the more memory it uses.

```
my $proc = Proc.new(:in);
$proc.spawn('cat');
while True {
  $proc.in.write("abc".encode);
}
```

## Environment
* Operating system: Ubuntu 20.04
* Compiler version (`perl6 -v` or `raku -v`):
```
This is Rakudo version 2020.07 built on MoarVM version 2020.07
implementing Raku 6.d.
```

## Diagnosis

I believe the problem is due to how promises are used internally in `Proc`.

The following code is roughly equivalent to how `Proc` sets up its promises (see the `$in === True` branch of `Proc::BUILD`) and also demonstrates the memory leak.

```
my $cur-promise = Promise.new;
$cur-promise.keep(True);
my $proc = Proc::Async.new('cat', :w);
$proc.start;
while True {
  $cur-promise = $cur-promise.then({ await $proc.write("abc".encode); });
}
```

However, the following (which removes the promise) does **not** have the memory leak.

```
my $cur-promise = Promise.new;
$cur-promise.keep(True);
my $proc = Proc::Async.new('cat', :w);
$proc.start;
while True {
  await $proc.write("abc".encode);
}
```

## Workaround

Until this memory leak is fixed, using `Proc::Async` directly can be used as a workaround for this issue.

## Related Issues

Note that while issue #3844 superficially looks related, that issue is dealing with the STDOUT and STDERR of a `Proc` while the current issue deals with STDIN.  Thus, I don't think it is actually related.

**Comment by AlexDaniel** [comment]
Isn't it simply that the while loop is producing the data a bit too fast? That is, rakudo just can't keep up writing to the Proc and the queue is growing.

**Comment by adamsmd** [comment]
@AlexDaniel I had not thought of that as a possible cause.  In order to test your theory, I tried both of the following versions, which I believe prevent data being written too fast.  They both still showed memory growth although the later program grows about 10 times slower.

(Also, the original application that led me to this bug also reads from the `Proc` after every write is a way that I think makes it impossible for data to be written too fast.)

Finally, note that I ran all these programs with `stdout` redirected to `/dev/null` so actual I/O time should be minimal.  It is possible that if someone doesn't do that redirect, I/O delays may cause other behaviors.

```
my $cur-promise = Promise.new;
$cur-promise.keep(True);
my $proc = Proc::Async.new('cat', :w);
$proc.start;
while True {
  $cur-promise = $cur-promise.then({ await $proc.write("abc".encode); });
  await $cur-promise; # NOTE: This is indented to block until the I/O finishes, but I am not familiar with it, so I might be using it wrong
}
```

```
my $cur-promise = Promise.new;
$cur-promise.keep(True);
my $proc = Proc::Async.new('cat', :w);
$proc.start;
while True {
  for 1..1000 {
    $cur-promise = $cur-promise.then({ await $proc.write("abc".encode); });
  }
  sleep 1;
}
```

**Comment by AlexDaniel** [comment]
I tried the last snippet (the one with `sleep 1`), and it does indeed leak pretty fast. However, it probably has nothing to do with Procs, and it's just the result of the weird stuff in the code with promises. The graph is rather interesting though:

![image](https://user-images.githubusercontent.com/5507503/90829521-911d7a00-e348-11ea-8bd0-9a2aad0f2bc1.png)

But a simple `await` does not leak, so it's not an issue in Procs I think.

**Comment by adamsmd** [comment]
When you say "a simple `await` does not leak", do you mean the snippet with `await $cur-promise`?  I double-checked it, and after 6 minutes, it had consumed 12GB (both virtual and resident memory).  Maybe there is a platform or version difference?

**Comment by AlexDaniel** [comment]
@adamsmd No, I mean `await $proc.write("abc".encode)`. You say “However, the following (which removes the promise)” but no promise is removed. `.write` is still returning a promise (otherwise what are you awaiting for?), and this way there's no memory leak that I was able to see. It is possible that creating new promises in interesting ways can leak, but that's basically my point – the issue is not in Proc::Async itself.

**Comment by adamsmd** [comment]
I agree `Proc::Async` is not at fault and that `await $proc.write("abc".encode)` does not leak.  Also, by "which removes the promise", I meant that it does not use a separate promise (though `$proc.write` returns a promise that the `await` waits on).

However, there may be an issue with `Proc` (not `Proc::Async`).  In particular, the `$cur-promise = $cur-promise.then` example is basically how `Proc` is implemented in terms of `Proc::Async` and promises: https://github.com/rakudo/rakudo/blob/aa1b3ed32f87580b2b03eac95a774ee67c5df7b8/src/core.c/Proc.pm6#L29-L42

I am not familiar enough with promises to know whether using them this way that should consume memory.

If using promises this way **should** keep consuming memory, then `Proc` is at fault since the `$cur-promise = $cur-promise.then` example is basically how `Proc` is implemented in terms of `Proc::Async` and promises.

If using promises this way **should not** keep consuming memory, then I agree that promises are at fault not `Proc`.

Finally, at your prompting, I found the following example, which demonstrates the problem without using `Proc` or `Proc::Async`.  (Though, again, I don't know whether using promises this way should consume memory.)

```
my $cur-promise = Promise.new;
$cur-promise.keep(True);
while True {
  $cur-promise = $cur-promise.then({ print '.' });
  await $cur-promise;
}
```
